main.c

```diff
@@ -1135,8 +1135,34 @@ static void term_in(int fd, short mask, void *data) {
 	state.run_display = false;
 }
 
-int main(int argc, char **argv) {
+// Check for --debug 'early' we also apply the correct loglevel
+// to the forked child, without having to first proces all of the
+// configuration (including from file) before forking and (in the
+// case of the shadow backend) dropping privileges
+void log_init(int argc, char **argv) {
+	static struct option long_options[] = {
+		{"debug", no_argument, NULL, 'd'},
+        {0, 0, 0, 0}
+    };
+    int c;
+	optind = 1;
+    while (1) {
+		int opt_idx = 0;
+		c = getopt_long(argc, argv, "d", long_options, &opt_idx);
```

Comment:
It looks like this PR broke option parsing for swaylock on my system

```suggestion
		c = getopt_long(argc, argv, ":d", long_options, &opt_idx);
```

Without the above change, swaylock still does not properly parse arguments unless `--debug` is included in the argument list somewhere, and even then it still prints a confusing error if `--debug` is not the very first argument

With the above change it does not print an error if `--debug` is not included and it does properly parses the other arguments again

From the getopt manual page:

> *  If  the first character (following any optional '+' or '-' described above) of optstring is a colon (':'), then getopt() likewise does not print an error message.  In addition, it returns ':' instead of '?' to indicate a missing option argument.  This allows the caller to distinguish the two different types of errors.

Reply:
Ugh, I'm so sorry about that. I swear I tested a number of variations, but obviously not sufficiently. I'll get on testing if adding `:` indeed seems sufficient when I try a number of different invocation variations, and then PR.

Reply:
No worries! I'm actually starting to notice still odd behavior even with this change, so doing `getopt_long` twice seems more of a beast than I thought

Reply:
Specifically I'm still seeing stuff like this, but only if `--debug` is not the first arg 🤔 (same behavior when it is excluded too):

```
[I] lily@bina ~/s/swaylock (fix/no-debug)> build/swaylock -c 000000 --bs-hl-color 333333 --debug
2023-01-25 11:19:53 - [main.c:1214] Parsing CLI Args
2023-01-25 11:19:53 - [main.c:39] Invalid color --bs-hl-color, defaulting to 0xFFFFFFFF
```

Reply:
> **By default, getopt() permutes the contents of argv as it scans**, so that eventually all the nonoptions are at the end.  Two other scanning modes are  also  implemented.  If the first character of optstring is '+' or the environment variable POSIXLY_CORRECT is set, then option processing stops as soon as a nonoption argument is encountered.  If '+' is not the first character of optstring, it is treated as a normal option.  If POSIXLY_CORRECT behaviour is required in this case optstring will contain two '+' symbols.  If the first character of optstring is '-', then each nonoption argv-element is handled as if it were the argument of an option with character code 1.

Ugh, I guess we need `-` in front of it too maybe

Reply:
ok, so it seems the first time around (in `log_init`) it actually mutates `argv` because there it doesn't know `-c` should take an argument. That seems like borderline a bug in getopt tbh... but I guess we should work around it in swaylock anyway.

Reply:
oh you beat me to it, I should refresh more often :laughing:

Reply:
Sweet! I opened #279